### PR TITLE
Dock panel

### PR DESCRIPTION
**Pull request by ajtudela**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | |
| Primary OS tested on | Ubuntu 22.04 |
| Robotic platform tested on | Morphia robot / Gazebo |
| Does this PR contain AI generated software? | No |

---
New panel for docking / undocking action, similar to Nav2 panel:
* Goal status and feedback.
* Error codes.
* Docking action its 1000% configurable: dock id, pose and type; and flag for nav to stage.

docking panel

[panel_docking.webm](https://github.com/ros-navigation/navigation2/assets/4681902/d8f61647-fdf4-4567-bd1d-6965cf670545)

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Very cool, I love this! Can we get this in the Nav2 default rviz config? Maybe on the right side of the window since the left side is getting pretty crowded. Or I suppose I wouldn't be opposed to rearranging the panel setup in rviz to be less "default" for a Nav2 specific layout. Perhaps:
- Move the displays panel (where you toggle/pick stuff to visualize) to the right top, with the views on the right bottom
- Move the Nav2 main panel to the top left, Nav2 docking center left, Nav2 controller selector bottom left

But just a suggestion! If we're going to have more of these plugins, its worth optimizing for them now!

File: nav2_rviz_plugins/src/docking_panel.cpp [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+    goal_msg.dock_pose.header.frame_id = "map";
+    goal_msg.dock_pose.header.stamp = client_node_->now();
+    goal_msg.dock_pose.pose.position.x = dock_pose_x_->text().toDouble();
+    goal_msg.dock_pose.pose.position.y = dock_pose_y_->text().toDouble();
+    goal_msg.dock_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+      dock_pose_yaw_->text().toDouble());
+    goal_msg.dock_type = combo_box->currentText().toStdString();
+
+    RCLCPP_INFO(
+      client_node_->get_logger(), "DockRobot will be called using dock pose: (%f, %f)",
```
Comment (SteveMacenski) [inline nav2_rviz_plugins/src/docking_panel.cpp:244]: Add the type here in the log

File: nav2_rviz_plugins/src/docking_panel.cpp [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+      RCLCPP_ERROR(client_node_->get_logger(), "Dock pose is empty.");
+      return;
+    }
+    goal_msg.dock_pose.header.frame_id = "map";
+    goal_msg.dock_pose.header.stamp = client_node_->now();
+    goal_msg.dock_pose.pose.position.x = dock_pose_x_->text().toDouble();
+    goal_msg.dock_pose.pose.position.y = dock_pose_y_->text().toDouble();
+    goal_msg.dock_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+      dock_pose_yaw_->text().toDouble());
+    goal_msg.dock_type = combo_box->currentText().toStdString();
```
Comment (SteveMacenski) [inline nav2_rviz_plugins/src/docking_panel.cpp:241]: How is the combo box populated with dock plugins?

Comment (ajtudela) [reply nav2_rviz_plugins/src/docking_panel.cpp:241]: The combobox is filled using the same ``pluginLoader`` function as the Selector: https://github.com/ros-navigation/navigation2/pull/4458/files#diff-e29dbc8bae4e148962b55e5b26da6a96d5dfca9fdb8223724abca2c463f3dd11R37

**Comment by ajtudela** [comment]
I personally use a config similar to this: 
rviz

**Comment by ajtudela** [comment]
So fixing the alignment for the docking panel, also fixed the alignment for the Nav2 panel!

rviz2

File: nav2_rviz_plugins/src/docking_panel.cpp (hunk 1 above)
Comment (ajtudela) [reply nav2_rviz_plugins/src/docking_panel.cpp]: Done

**Comment by SteveMacenski** [comment]
Great! Let me know when you want me to look at it again!

Can you try a layout where all the Nav2 stuff is on one side? It would be nice if it were grouped together!

**Comment by ajtudela** [comment]
> Great! Let me know when you want me to look at it again! This is really quick turn around on this tool and its unquestionably valuable to have, thanks!
> 
> Can you try a layout where all the Nav2 stuff is on one side? It would be nice if it were grouped together!

You can review it now. I've fixed the log and the layout

**Comment by SteveMacenski** [comment]
Can you also push the updated rviz config?

**Comment by ajtudela** [comment]
> 
> Can you try a layout where all the Nav2 stuff is on one side? It would be nice if it were grouped together!

The only option is to merge the panels in "tabs". You can see them at the bottom.
Captura desde 2024-06-21 20-25-52

**Comment by SteveMacenski** [comment]
> The only option is to merge the panels in "tabs". You can see them at the bottom.

Can't you move the top Displays panel to the right hand side?

**Comment by ajtudela** [comment]
> > The only option is to merge the panels in "tabs". You can see them at the bottom.
> 
> Can't you move the top Displays panel to the right hand side where the Views are?

After a bit of struggle, if have this:
Captura desde 2024-06-21 20-39-34

**Comment by SteveMacenski** [comment]
Eh ok, you've convinced me about https://github.com/ros-navigation/navigation2/pull/4458#issuecomment-2183217347 can you push that one? I'll test and merge

**Comment by ajtudela** [comment]
Done!

**Comment by SteveMacenski** [comment]
Weird... basically all the tests failed. Trying again!

**Comment by ajtudela** [comment]
> Weird... basically all the tests failed. Trying again!

Yeah, I thought CI was failing again. I made a rebase and now the tests passed.

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Thanks! This looks great! Amazing tool to have now for users!

**Comment by SteveMacenski** [comment]
@ajtudela I'm running into a problem while using the panel. It doesn't appear that the terminal state (i.e. action completed) is updating the state machine properly. When docking, it works fine but then never exits the "active action" state (i.e. keeps the dock/undock button greyed out and the cancel button active) even though the action was long completed. That prevents us from being able to undock or redock

**Comment by ajtudela** [comment]
It's weird, I saw the code of the new non-charging dock type and it doesn't change the code of the docking server. The status of the buttons and text only depends on the goal status actions. Maybe these callbacks are not called when the action is completed?

https://github.com/ros-navigation/navigation2/blob/c2e4aabdab5a4d3ebbf011c9e28d00ce3e27a4c3/nav2_rviz_plugins/src/docking_panel.cpp#L152-L181

I'll check it out as soon as I get back to the lab.

**Comment by ajtudela** [comment]
> @ajtudela I'm running into a problem while using the panel developing #4405 on `main`. It doesn't appear that the terminal state (i.e. action completed) is updating the state machine properly. When docking, it works fine but then never exits the "active action" state (i.e. keeps the dock/undock button greyed out and the cancel button active) even though the action was long completed. That prevents us from being able to undock or redock

I've tried several times to perform dock/undock actions using the new non-charging dock successfully using the dock panel . Can you give me more details about your system so I can try to reproduce the bug?

**Comment by SteveMacenski** [comment]
That is weird, because I saw this consistently. I used the `main` branch and the provided `nav2_bringup` launch files so the server bringup / branch should be the same between us (nothing custom). Then using the rviz tool which interacts with the docking server fine (i.e. it docks) but then the rviz panel state seems to miss the results or something

**Pull request merged**